File: app/schemas/game.py
from datetime import datetime

from pydantic import BaseModel, Field

from app.schemas.common import BaseSchema


class ResourceAmount(BaseSchema):
    resource: str
    amount: float


class ModuleView(BaseSchema):
    module_key: str
    level: int
    is_active: bool


class ModuleDefinitionView(BaseSchema):
    key: str
    name: str
    description: str
    category: str
    max_level: int
    base_cost: dict[str, float]
    upgrade_cost_growth: float
    energy_delta: float
    throughput_delta: float
    crew_requirement: int
    sort_order: int


class StationView(BaseSchema):
    id: str
    name: str
    level: int
    specialization: str
    throughput: float
    efficiency: float
    stability: float
    reputation: float
    bottlenecks: list[str]
    recommended_actions: list[str]
    inventories: list[ResourceAmount]
    modules: list[ModuleView]
    module_catalog: list[ModuleDefinitionView]
    last_processed_at: datetime


class StationRenameRequest(BaseModel):
    name: str = Field(min_length=3, max_length=60)


class ModuleActionRequest(BaseModel):
    module_key: str


class PolicyRequest(BaseModel):
    key: str
    value: str


class ReportView(BaseSchema):
    id: str
    started_at: datetime
    ended_at: datetime
    summary: dict[str, object]
    claimed_at: datetime | None


class MarketTradeRequest(BaseModel):
    resource: str
    quantity: float = Field(gt=0, le=9999)


class MarketStateView(BaseSchema):
    resource: str
    price: float
    trend: float
    history: list[float]


class ContractCreateRequest(BaseModel):
    title: str = Field(min_length=3, max_length=120)
    contract_type: str
    resource: str
    quantity: float = Field(gt=0)
    reward_credits: float = Field(gt=0)


class TransferRequest(BaseModel):
    target_station_id: str
    resource: str
    amount: float = Field(gt=0)
    note: str = Field(default="", max_length=120)


class MetaUpgradeView(BaseSchema):
    key: str
    name: str
    description: str
    base_cost: int
    max_level: int
    effect_type: str
    effect_value: float
    current_level: int


class MetaPurchaseRequest(BaseModel):
    key: str


class NotificationView(BaseSchema):
    id: str
    type: str
    title: str
    message: str
    payload: dict[str, object]
    read_at: datetime | None


class ChatMessageView(BaseSchema):
    id: str
    sender_user_id: str
    sender_username: str
    recipient_user_id: str | None
    body: str
    created_at: datetime


class ChatThreadView(BaseSchema):
    user_id: str
    username: str
    station_name: str | None
    last_message: str | None
    last_message_at: datetime | None
    unread_count: int


class ChatSendRequest(BaseModel):
    body: str = Field(min_length=1, max_length=300)


class SectorPlayerView(BaseSchema):
    station_id: str
    station_name: str
    owner_username: str
    specialization: str
    level: int
    reputation: float


class SectorSnapshot(BaseSchema):
    sector_id: str
    sector_name: str
    market_mode: str
    market_mood: str
    players: list[SectorPlayerView]
    events: list[dict[str, object]]

File: frontend/src/types/game.ts
export type ResourceAmount = { resource: string; amount: number }
export type ModuleView = { module_key: string; level: number; is_active: boolean }
export type ModuleDefinition = {
  key: string
  name: string
  description: string
  category: string
  max_level: number
  base_cost: Record<string, number>
  upgrade_cost_growth: number
  energy_delta: number
  throughput_delta: number
  crew_requirement: number
  sort_order: number
}
export type StationView = {
  id: string
  name: string
  level: number
  specialization: string
  throughput: number
  efficiency: number
  stability: number
  reputation: number
  bottlenecks: string[]
  recommended_actions: string[]
  inventories: ResourceAmount[]
  modules: ModuleView[]
  module_catalog: ModuleDefinition[]
  last_processed_at: string
}

export type ReportView = {
  id: string
  started_at: string
  ended_at: string
  summary: Record<string, unknown>
  claimed_at: string | null
}

export type MarketRow = { resource: string; price: number; trend: number; history: number[] }
export type ContractRow = {
  id: string
  title: string
  contract_type: string
  resource: string
  quantity: number
  reward_credits: number
  reward_reputation: number
  status: string
  source: string
  issuer_station_id: string | null
  taker_station_id: string | null
  expires_at: string
}
export type MetaUpgrade = {
  key: string
  name: string
  description: string
  base_cost: number
  max_level: number
  effect_type: string
  effect_value: number
  current_level: number
}
export type SectorSnapshot = {
  sector_id: string
  sector_name: string
  market_mode: string
  market_mood: string
  players: {
    station_id: string
    station_name: string
    owner_username: string
    specialization: string
    level: number
    reputation: number
  }[]
  events: { id: string; title: string; description: string; ends_at: string }[]
}
export type NotificationRow = {
  id: string
  type: string
  title: string
  message: string
  payload: Record<string, unknown>
  read_at: string | null
}

export type LiveDashboardSnapshot = {
  station: StationView
  reports: ReportView[]
  market: MarketRow[]
  npc_contracts: ContractRow[]
  npc_contract_visibility: number
}

export type ChatMessageRow = {
  id: string
  sender_user_id: string
  sender_username: string
  recipient_user_id: string | null
  body: string
  created_at: string
}

export type ChatThreadRow = {
  user_id: string
  username: string
  station_name: string | null
  last_message: string | null
  last_message_at: string | null
  unread_count: number
}

export type AdminContentType =
  | 'resource'
  | 'module'
  | 'event'
  | 'contract_template'
  | 'meta_upgrade'
  | 'specialization'

export type AdminContentItem = {
  id: number
  content_type: AdminContentType
  key: string
  display_name: string
  source_kind: string
  status: string
  tags: string[]
  current_revision_id: number | null
  published_revision_id: number | null
  updated_by: string | null
  updated_at: string
  payload?: Record<string, unknown> | null
}

export type AdminContentRevision = {
  id: number
  version: number
  change_summary: string
  is_published: boolean
  published_at: string | null
  author_user_id: string | null
  payload_json: Record<string, unknown>
  created_at: string
}

export type AdminBalanceItem = {
  id: number
  key: string
  category: string
  scope: string
  enabled: boolean
  value_json: Record<string, unknown>
  current_revision_id: number | null
  published_revision_id: number | null
  updated_at: string
}

export type AdminBalanceRevision = {
  id: number
  version: number
  change_summary: string
  is_published: boolean
  published_at: string | null
  author_user_id: string | null
  value_json: Record<string, unknown>
  created_at: string
}

export type AdminAuditLogRow = {
  id: number
  actor_user_id: string | null
  action_type: string
  target_type: string
  target_id: string
  summary: string
  metadata_json: Record<string, unknown>
  created_at: string
}

export type AdminUserRow = {
  id: string
  username: string
  email: string
  roles: string[]
}

export type AdminPlayerSummaryRow = {
  station_id: string
  owner_user_id: string
  username: string
  email: string
  is_active: boolean
  station_name: string
  specialization: string
  level: number
  throughput: number
  efficiency: number
  stability: number
  reputation: number
  updated_at: string
}

export type AdminPlayerDetailRow = AdminPlayerSummaryRow & {
  public_notes: string
  inventories: { resource: string; amount: number }[]
  modules: { module_key: string; level: number; is_active: boolean }[]
  last_processed_at: string
}

export type AdminAuthz = {
  user_id: string
  username: string
  roles: string[]
  permissions: string[]
}
